### ISSUE TITLE
Non-deterministic NaN in cho_factor

### ISSUE THREAD

**Issue by davisyoshida** [google/jax#15769]
### Description

The following code gives me NaNs about half the time, and a correct factorization the other half:
```python
import jax
import jax.numpy as jnp
mat = jax.random.normal(jax.random.PRNGKey(0), (4096, 4096))    
mat = mat @ mat.T    
mat /= jnp.max(jnp.abs(mat))    
cho, upper = jax.scipy.linalg.cho_factor(mat)    
print(cho)
```

This only occurs on GPU, not on CPU.

### What jax/jaxlib version are you using?
0.4.8/0.4.7

### Which accelerator(s) are you using?

GPU

### NVIDIA GPU info

(This was done with the Quadro RTX 6000)
Wed Apr 26 23:33:44 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.41.03              Driver Version: 530.41.03    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce GTX 1070 Ti      Off| 00000000:01:00.0 Off |                  N/A |
| 29%   46C    P0               38W / 180W|      0MiB /  8192MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  Quadro RTX 6000                 Off| 00000000:03:00.0 Off |                  Off |
| 33%   42C    P0               27W / 260W|      0MiB / 24576MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

**Comment by hawkinsp** [comment]
Can you confirm what CUDA version you have installed? (nvidia-smi reports the CUDA version of the driver, but I'm curious what the CUDA version of the *libraries* you have installed is.) How did you install JAX?

I can't reproduce this on a Colab T4 GPU using CUDA 12, so something must be different about my setup and yours.

**Comment by davisyoshida** [comment]
I installed jax with the command:
```bash
$ pip install --upgrade "jax[cuda12_pip]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html
```

I have CUDA 12.1.1 and CUDNN 8.8.0 installed.

**Comment by davisyoshida** [comment]
Looks like it. I upgraded to jax/jaxlib version 0.4.13, ran the script 50 times, and got NaNs 37/50 times.

**Comment by hawkinsp** [comment]
Thanks. Hmm. We'll have to find a GPU it reproduces on.

**Comment by hawkinsp** [comment]
Yes, it's certainly possible that a matrix that is almost PSD might act differently on different GPUs, or even perhaps nondeterministically on the same GPU due to concurrency changing order of operations. I'm still a bit surprised that's it's nondeterministic.

**Comment by hawkinsp** [comment]
I was able to reproduce this problem on V100. By printing hashes of the various arrays in the program, I found that it's actually the matrix multiplication `mat @ mat.T` that is nondeterministic. Different matmul algorithms produce slightly different results, and that's enough to prevent the Cholesky decomposition from working in some cases.

In principle one should be able fix this by setting `XLA_FLAGS=--xla_gpu_deterministic_ops=1`, but I tried that and it still doesn't work. I filed an XLA:GPU bug (https://github.com/openxla/xla/issues/6796).
